File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        int sArray[] = new int[26];
        int pArray[] = new int[26];

        if(p.length() > s.length()) return new ArrayList<>(0);

        for(char c: p.toCharArray()) {
            pArray[c-'a']++;
        }

        int k = p.length();
        for(int i=0; i<k; i++) {
            char c = s.charAt(i);
            sArray[c-'a']++;
        }
        if(Arrays.equals(pArray, sArray)) {
                res.add(0);
            }

        for(int i=k; i<s.length(); i++) {
            sArray[s.charAt(i-k) - 'a']--;
            sArray[s.charAt(i)-'a']++;
            if(Arrays.equals(pArray, sArray)) {
                res.add(i-k+1);
            }

        }

        // for(int i=0; i<26; i++) {
        //     System.out.println(sArray[i]+" "+pArray[i]);
        // }

        return res;
    }
}

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        pArray = [0]*26
        sArray = [0]*26

        if len(s) < len(p):
            return []

        res = []

        for c in p:
            pArray[ord(c)-97]+=1

        k = len(p)

        

        for i in range(0, len(p)):
            sArray[ord(s[i]) - 97] += 1

        if sArray == pArray:
            res.append(0)

        for i in range(k, len(s)):
            sArray[ord(s[i])-97]+=1
            sArray[ord(s[i-k])-97]-=1
            if sArray == pArray:
                res.append(i-k+1)

        return res



        